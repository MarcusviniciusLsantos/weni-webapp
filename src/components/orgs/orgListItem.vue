<template>
  <unnnic-card-company
    :title="name"
    :tag="$t(`billing.${plan}.title`)"
    :description="description"
    :join-label="$t('orgs.join')"
    :options="options"
    :members="displayMembers"
    :members-description="
      remainingMembers ? $tc('orgs.remaining_members', remainingMembers) : null
    "
    @join="onSelectOrg"
  />
</template>

<script>
import image from '../../../src/assets/default-avatar.svg';

const ORG_ROLE_CONTRIBUTOR = 2;
const ORG_ROLE_ADMIN = 3;
const ORG_ROLE_FINANCIAL = 4;

export default {
  name: 'OrgListItem',
  props: {
    name: {
      type: String,
      default: null,
    },
    description: {
      type: String,
      default: null,
    },
    members: {
      type: Array,
      default: () => [],
    },
    canEdit: {
      type: Boolean,
      default: false,
    },
    canEditBilling: {
      type: Boolean,
      default: false,
    },
    role: {
      type: Number,
    },
    org: {
      type: Object,
    },
  },
  data() {
    return {
      highlighted: false,
      confirmOrganizationName: '',
    };
  },
  computed: {
    plan() {
      return this.org.organization_billing?.plan;
    },

    displayMembers() {
      if (!this.members) return [];

      return this.members.map((member) => ({
        name: this.getName(member),
        photo: member.photo_user || image,
      }));
    },
    remainingMembers() {
      if (!this.members || this.members.length <= 4) return 0;
      return this.members.length - 4;
    },

    options() {
      const billingArray = [
        {
          icon: 'currency-dollar-circle-1',
          title: this.$t('orgs.billing'),
          click: this.onSelectBilling,
        },
      ];

      if (this.role === ORG_ROLE_CONTRIBUTOR) {
        return [
          {
            icon: 'view-1-1',
            title: this.$t('orgs.view_members'),
            click: this.onView,
          },
        ];
      } else if (this.role === ORG_ROLE_FINANCIAL) {
        return billingArray;
      } else if (this.role === ORG_ROLE_ADMIN) {
        return [
          {
            icon: 'single-neutral-actions-1',
            title: this.$t('orgs.manage_members'),
            click: this.onManage,
          },
        ]
          .concat(billingArray)
          .concat([
            {
              icon: 'cog-1',
              title: this.$t('orgs.config'),
              click: this.onEdit,
            },
            {
              icon: 'delete-1-1',
              title: this.$t('orgs.delete.title'),
              click: this.openDeleteModal,
              scheme: 'feedback-red',
            },
          ]);
      }

      return [];
    },
  },
  methods: {
    openDeleteModal() {
      this.$emit('open-delete-confirmation');
    },

    getName(user) {
      const name = [user.first_name, user.last_name]
        .filter((name) => name)
        .join(' ');
      return name ? name : user.username;
    },

    onSelectOrg() {
      if (this.role === ORG_ROLE_FINANCIAL) {
        this.onSelectBilling();
      } else {
        this.$emit('select');
      }
    },
    onSelectBilling() {
      this.$emit('billing');
    },
    onEdit() {
      this.$emit('edit');
    },
    onManage() {
      this.$emit('manage');
    },
    onView() {
      this.$emit('view');
    },
  },
};
</script>
