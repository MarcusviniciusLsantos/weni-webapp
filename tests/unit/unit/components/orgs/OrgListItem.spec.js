import { shallowMount, createLocalVue } from '@vue/test-utils';
import orgListItem from '@/components/orgs/orgListItem.vue';
import i18n from '@/utils/plugins/i18n';
import { org } from '../../../__mocks__';

const localVue = createLocalVue();

describe('orgListItem.vue', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(orgListItem, {
      localVue,
      i18n,
      propsData: {
        steps: 1,
        names: [],
        current: 0,
        org,
      },
      mocks: {
        $t: () => 'some specific text',
      },
      stubs: {
        UnnnicCardCompany: true,
      },
    });
  });

  it('renders a snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('getName()', () => {
    const response = wrapper.vm.getName({
      first_name: 'Test',
      last_name: 'Vue',
    });

    expect(response).toBe('Test Vue');
  });

  describe('displayMembers()', () => {
    it('Test when dont have members passing by props', async () => {
      await wrapper.setProps({ members: null });
      const response = wrapper.vm.displayMembers;

      expect(response).toStrictEqual([]);
    });

    it('Test when have members passing by props', async () => {
      const members = [
        { first_name: 'Test', last_name: 'Vue', photo_user: 'vue' },
      ];
      await wrapper.setProps({ members });
      expect(wrapper.vm.displayMembers).toStrictEqual([
        {
          name: 'Test Vue',
          photo: 'vue',
        },
      ]);
    });
  });

  describe('remainingMembers()', () => {
    it('When have more or equal to 4 members', async () => {
      const members = [1, 2, 3, 4, 5];
      await wrapper.setProps({ members });

      const response = wrapper.vm.remainingMembers;
      expect(response).toStrictEqual(members.length - 3);
    });

    it('When have less then 4 members', async () => {
      const members = [1, 2, 3];
      await wrapper.setProps({ members });

      const response = wrapper.vm.remainingMembers;
      expect(response).toStrictEqual(0);
    });
  });

  describe('options()', () => {
    it('When can edit', async () => {
      await wrapper.setProps({ role: 3 });

      const response = wrapper.vm.options;
      expect(response.length).toStrictEqual(4);
    });

    it('When cannot edit', async () => {
      await wrapper.setProps({ role: 2 });
      const optionsRole2 = wrapper.vm.options;
      expect(optionsRole2.length).toStrictEqual(1);

      await wrapper.setProps({ role: 4 });
      const optionsRole4 = wrapper.vm.options;
      expect(optionsRole4.length).toStrictEqual(1);
    });
  });

  it('openDeleteModal()', () => {
    wrapper.vm.openDeleteModal();
    expect(wrapper.emitted('open-delete-confirmation')).toBeTruthy();
  });

  it('onSelectOrg()', () => {
    wrapper.vm.onSelectOrg();
    expect(wrapper.emitted('select')).toBeTruthy();
  });

  it('onEdit()', () => {
    wrapper.vm.onEdit();
    expect(wrapper.emitted('edit')).toBeTruthy();
  });

  it('onManage()', () => {
    wrapper.vm.onManage();
    expect(wrapper.emitted('manage')).toBeTruthy();
  });

  it('onView()', () => {
    wrapper.vm.onView();
    expect(wrapper.emitted('view')).toBeTruthy();
  });

  it('onSelectBilling()', () => {
    wrapper.vm.onSelectBilling();
    expect(wrapper.emitted('billing')).toBeTruthy();
  });
});
