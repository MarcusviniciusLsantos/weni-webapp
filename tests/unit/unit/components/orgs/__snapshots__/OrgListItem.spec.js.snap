// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`orgListItem.vue renders a snapshot 1`] = `<unnniccardcompany-stub tag="some specific text" joinlabel="some specific text" options="" members="" steps="1" names="" current="0"></unnniccardcompany-stub>`;
